### PR TITLE
MSBuild self hosts on .net core (work in progress)

### PR DESCRIPTION
**Pull request by cdmihai**
- use net core msbuild from build tools to build msbuild for net core
- use dotnet CLI to restore projects, build tools, and runtime dependencies
- use build tools to drive project restore, not our own mechanism
- remove override.targets
- remove LinuxDeployDependencies
- remove compilers from packages since they are now coming via runtime dependencies
- replace packages.config with project.json

### CODE REVIEW

**Comment by cdmihai** [comment]
@radical 

I might have busted the mono target build with this PR. Unfortunately I do not know the details of the mono build to fix it. Can you please test if this still builds for mono, and then if not, send a subsequent PR?

I am trying to make the msbuild build scripts as small and possible by reusing as much as possible from the build tools package.

**Comment by cdmihai** [comment]
OSX CI is failing because HOME is not set for them. I'll need to set HOME for init-tools.sh as well.

File: src/.nuget/packageLoad.targets [hunk 1]
```diff
@@ -87,10 +54,14 @@
            AlwaysCreate="true"
            ForceTouch="true" />
 
+    <!-- Restore tools from packages/project.json -->
+    <Exec Condition="Exists('$(NuGetProjectJsonFile)')" 
+          Command="$(DnuCoreRestoreCommand) --packages $(ToolPackagesDir) &quot;$(NuGetProjectJsonFile)&quot;"/>
+
+<!--     <Exec Condition="Exists('$(NuGetProjectJsonFile)')"
```
Comment (cdmihai) [inline src/.nuget/packageLoad.targets]: remove this comment

File: build.cmd [hunk 2]
```diff
@@ -13,12 +13,17 @@ if not defined MSBUILDCUSTOMPATH (
...
+)
+
 echo ** MSBuild Path: %MSBUILDCUSTOMPATH%
 echo ** Building all sources
 
 :: Call MSBuild
-echo ** "%MSBUILDCUSTOMPATH%" "%~dp0build.proj" /verbosity:minimal /nodeReuse:false /fileloggerparameters:Verbosity=diag;LogFile="%~dp0msbuild.log" %*
-"%MSBUILDCUSTOMPATH%" "%~dp0build.proj" /verbosity:minimal /nodeReuse:false /fileloggerparameters:Verbosity=diag;LogFile="%~dp0msbuild.log" %*
+echo ** "%RUNTIMEHOST%" "%MSBUILDCUSTOMPATH%" "%~dp0build.proj" /verbosity:minimal /fileloggerparameters:Verbosity=diag;LogFile="%~dp0msbuild.log" %*
+"%RUNTIMEHOST%" "%MSBUILDCUSTOMPATH%" "%~dp0build.proj" /verbosity:minimal /fileloggerparameters:Verbosity=diag;LogFile="%~dp0msbuild.log" %*
```
Comment (rainersigwald) [inline build.cmd]: Does this work if you want to build with full-framework MSBuild and RUNTIMEHOST isn't defined (and then feeds in empty quotes)?

**Comment by AndyGerlicher** [comment]
LGTM. Please note what scenarios work and which don't yet and let's get this in so we can more forward off this.

**Comment by cdmihai** [comment]
Scenario build status:

- Windows
 - works: 
   - full framework msbuild -> Debug-NetCore 
    - .net core msbuild -> Debug-NetCore
 - does not work: 
   - full framework msbuild -> Debug-Mono (fails in the same way as xplat)
    - full framework msbuild -> Debug (compiles but some tests fail; fails in the same way as xplat)

- OSX
 - works:
   - .net core msbuild -> Debug-NetCore
    - mono msbuild -> Debug-NetCore
 - does not work:
   - mono msbuild -> Debug-Mono (the new PrereleaseResoveNugetPackageAssets task seems to resolve more references when the target framework is 4.5.1 than when the target framework is 4.6 on full framework msbuild-> Debug build. This crashes mcs)

**Comment by cdmihai** [comment]
I set cibuild.sh to use a mono host by default to make CI happy. Will need to change back to core clr when build tools updates their msbuild package.

**Comment by rainersigwald** [comment]
LGTM

**Comment by cdmihai** [comment]
@dotnet-bot test this please

**Comment by cdmihai** [comment]
@dotnet-bot test Ubuntu Build please

**Comment by cdmihai** [comment]
@dotnet-bot test this please

**Pull request merged**